### PR TITLE
add change [Admin] PIN method/cli

### PR DESCRIPTION
**Pull request by kilimnik**
I implemented my feature request #472.

### CODE REVIEW

**Comment by dainnilsson** [comment]
Please rebase this PR and target the `next` branch, rather than `main`, as that is where active development for the next substantial version of the project is happening.

Also, please take a look at the `access` subcommand for `piv`, as it implements similar commands. To the extent possible, please use the same option names and short forms as `piv access change-pin` does (Eg. `--pin, -P`, not `--old-pin, -o`).

Once rebased for `next` I'll take another look!

**Review by dainnilsson — CHANGES_REQUESTED** [review]
I'd like to have some tests for this functionality as well, probably in `tests/device/cli/test_openpgp.py`, something similar to https://github.com/Yubico/yubikey-manager/blob/main/tests/device/cli/piv/test_pin_puk.py

**Comment by dainnilsson** [comment]
Thinking some more about this functionality:
I don't think it makes sense to include it in a release unless we also have support for changing and using the "Reset Code". I'm not saying that *you* have to implement that feature, just that *someone* will have to before it makes sense to include it in the application. I can merge this PR (and the one for "signature PIN") into a separate branch for now, but until the "PIN management" feature is completed with that it won't go into `next` or `main`.

File: ykman/cli/openpgp.py [hunk 1]
```diff
@@ -183,6 +183,54 @@ def set_pin_retries(
             echo_default_pins()
 
 
+@access.command("change-pin")
+@click.option("-P", "--pin", help="Current PIN code.")
+@click.option("-n", "--new-pin", help="A new PIN.")
+@click.pass_context
+def change_pin(ctx, pin, new_pin):
+    """
+    Change PIN.
```
[the remark this answers is not in the document]
Comment (kilimnik) [reply ykman/cli/openpgp.py:192]: What about maximum length, I saw in the documentation for OpenPGP Smart Card spec, that the max length can be read via getDO. It is set to 63 on my YubiKeys, is that something I should note too, as it is possible for Yubico to change this

Also I am unure about cross platform compatibility, the spec says UTF-8

File: ykman/cli/openpgp.py [hunk 2]
```diff
@@ -183,6 +183,54 @@ def set_pin_retries(
...
+        new_pin_repeated = click_prompt("Repeat new PIN", hide_input=True)
+
+        if new_pin != new_pin_repeated:
+            cli_fail("PIN not correctly repeated.")
+
+    controller.change_pin(pin, new_pin)
+
+
+@access.command("change-admin")
+@click.option("-o", "--old-admin", help="Old Admin PIN for OpenPGP.")
```
[the remark this answers is not in the document]
Comment (kilimnik) [reply ykman/cli/openpgp.py:211]: I wrote the `for OpenPGP` part for consistency. See [here](https://github.com/Yubico/yubikey-manager/blob/77fe4e4a6c2a881819f765e91ce22c32d69bb9eb/ykman/cli/openpgp.py#L147)

[the remark this answers is not in the document]
Comment (dainnilsson) [reply ykman/cli/openpgp.py:211]: Good point, it looks like we're inconsistent in this regard already. Either is fine for now and I think we'll probably go over these at some point in the future and make sure we are more consistent.

File: ykman/cli/openpgp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dainnilsson) [reply ykman/cli/openpgp.py:192]: As far as I know, YubiKeys have always supported a max length of 127 and I think is unlikely to change that, so while the correct thing would be to read the DO, it's probably also fine to just hardcode this value. There's also the KDF feature which is done on the host side. If this is enabled, then the PIN lengths aren't really applicable in the same way.

Either way, this value is big enough that users are unlikely to run into this, so the upper bound isn't as important to state here as the minimum is. I think it should be safe to state that alphanumeric values are supported. Theoretically you could use more fancy unicode values, but you may run into issues when inputting this PIN on different platforms.

File: ykman/cli/openpgp.py [hunk 3]
```diff
@@ -183,6 +183,60 @@ def set_pin_retries(
...
 
+@access.command("change-pin")
+@click.option("-P", "--pin", help="Current PIN code.")
+@click.option("-n", "--new-pin", help="A new PIN.")
+@click.pass_context
+def change_pin(ctx, pin, new_pin):
+    """
+    Change PIN.
+
+    The PIN must be atleast 6 long, and supports any type of
```
Comment (dainnilsson) [inline ykman/cli/openpgp.py:194]: "at least" should be written as two words. "6 long" needs a unit. I would suggest "at least 6 characters long", but that repeats the word "characters" in the sentence. Maybe "The PIN has a minimum length of 6, and supports any type of alphanumeric characters."?

**Review by dainnilsson — CHANGES_REQUESTED** [review]
Looking good! I just have some minor corrections for the docstrings.

File: ykman/cli/openpgp.py [hunk 4]
```diff
@@ -183,6 +183,60 @@ def set_pin_retries(
...
+
+@access.command("change-admin-pin")
+@click.option("-a", "--admin-pin", help="Current Admin PIN.")
+@click.option("-n", "--new-admin-pin", help="New Admin PIN.")
+@click.pass_context
+def change_admin(ctx, admin_pin, new_admin_pin):
+    """
+    Change Admin PIN.
+
+    The Admin PIN has a minimum length of 6, and supports any type of
```
Comment (dainnilsson) [inline ykman/cli/openpgp.py:221]: I think you accidentally changed 8 to 6 here.

Comment (kilimnik) [reply ykman/cli/openpgp.py:221]: You're right will fix

**Review by dainnilsson — APPROVED** [review]

**Pull request closed without merging**